java-frontend/src/main/java/org/sonar/java/model/statement/CaseLabelTreeImpl.java

```diff
@@ -34,20 +35,25 @@
 
 public class CaseLabelTreeImpl extends JavaTree implements CaseLabelTree {
   private final InternalSyntaxToken caseOrDefaultKeyword;
-  @Nullable
-  private final ExpressionTree expression;
-  private final InternalSyntaxToken colonToken;
+  private final List<ExpressionTree> expressions;
+  private final Kind kind;
+  private final InternalSyntaxToken colonOrArrowToken;
+
+  public CaseLabelTreeImpl(InternalSyntaxToken caseOrDefaultKeyword, @Nullable ExpressionTree expression, InternalSyntaxToken colonOrArrowToken) {
+    this(caseOrDefaultKeyword, Collections.singletonList(expression), Kind.CASE_LABEL_COLON, colonOrArrowToken);
```

Comment:
I don't like the fact of having two distinct Kinds associated with the same Tree. Moreover, it introduces a breaking change in API, which I REALLY don't like.

What I propose is to introduce a new method, for instance `isFallThrough()` which would be equivalent in term of functionality, and the proper javadoc detailing the difference between "colon" and "arrow" cases. The CASE_LABEL then stays the same, and we only enrich the API with some new methods, and a deprecated one.